### DISCUSSION TITLE
Putting on the "xUnit Documentation Person" hat

### DISCUSSION THREAD

**Discussion by AlanMcBee** [xunit/xunit#2906] [General]
This is me introducing myself to the xUnit community and getting started on my pet project: bringing the overall documentation for xUnit to a higher quality than where it's at now. I don't have many details yet on what all that will mean, but at the start it will mean bringing back most (all?) of the reference docs that are currently absent from the documentation website. I have big ideas, but I'm new to being a long-term collaborator, so I'm expecting to just learn the ropes for now. That means going over all the issues and discussions related to documentation and digesting all of that before committing to anything beyond bringing back the reference docs.

**Comment by AlanMcBee** [comment]
I'm reviewing many document generator tools.

- [Atomineer](https://www.atomineerutils.com/)

- [ClickHelp](https://clickhelp.com/)

- [Codartis](https://codartis.com/)

- [DocFX](https://dotnet.github.io/docfx/)

- [Doc-O-Matic](https://www.doc-o-matic.com/en/index.html)

- [Document! X](https://documentation.contiem.com/HelpAuthoringTools/overview.html)

- [Doxie](https://github.com/gordon-matt/Doxie)

- [Doxygen](https://www.doxygen.nl/index.html)

- [Dr.Explain](https://www.drexplain.com/)

- [GhostDoc](https://submain.com/ghostdoc/)

- [Help & Manual](https://www.helpandmanual.com/)

- [Help+Manual](https://www.helpandmanual.com/)

- [HelpNDoc](https://www.helpndoc.com/)

- [HelpStudio](https://contiem.com/software/authoring/helpstudio/)

- [Hugo](https://gohugo.io/documentation/)
- [Live Documenter](https://livedocumenter.barryjones.me.uk/)

- [MkDocs](https://www.mkdocs.org/)

- [Monodoc](https://www.mono-project.com/docs/tools+libraries/tools/monodoc/)

- [Natural Docs](https://naturaldocs.org/)

- [Quick Diagram Tool for C#](https://github.com/realvizu/QuickDiagram) (now Codartis)

- [Sandcastle](https://github.com/EWSoftware/SHFB)

- [Sphinx](https://www.sphinx-doc.org/en/master/)

- [VSdocman](https://www.helixoft.com/vsdocman/overview.html)

There are more: https://en.wikipedia.org/wiki/Comparison_of_documentation_generators

Some are not free, so I'm not seriously considering them.

**DocFX** is Microsoft's, and seems like the most natural fit so far. It has a large number of features, is actively maintained, integrates well with the tools.

I'm not sure right now how much work would be required to get something like a responsive site output so that the docs could be called up on a mobile device, for example, but also having a desktop site that is capable of delivering a rich experience.

How important is it that the generation process includes good integration with IDE tools (VS, VSCode, ... Rider? Eclipse?)?

**Reply by bradwilson** [reply]
Welcome! 🎉 And thank you!

It is worth making some requirements/limitations clear up front:

* Whatever process we land on needs to work in the context of GitHub Pages, which is statically generated HTML via Jekyll templates. There is no runtime server component available here, so any interactivity needs to be done via client-side JavaScript. There is some "build time" automation that can be done (for example, [we generate the lists of available doc pages for the analyzers via Jekyll](https://github.com/xunit/xunit/blob/2d5ea01e7acc9b3090d8686e69aaaaf3e396d0da/xunit.analyzers/rules/index.html)). For more information on the templating capabilities of Jekyll, see: https://jekyllrb.com/
* The CI build process runs on Windows. The binary signing process that we use for CI builds requires Windows, and is unlikely to ever work on a non-Windows OS. Supporting other OSes falls under the "nice to have but not necessary" category.

> Some are not free, so I'm not seriously considering them.

Agreed. If they're not free for OSS projects, then they should be disqualified.

> DocFX is Microsoft's, and seems like the most natural fit so far.

We previously used Sandcastle when we were generating CHM files for the v1 releases, back in the day when it was owned by Microsoft. We stopped shipping those probably right around the time we switched to NuGet packages as they fell out of favor (I'm not even sure whether the CHM viewer is still part of modern versions of Windows). In point of fact, now that CodePlex is gone, it's unlikely any CHM files still exist in a known location. 🤷🏼 

> I'm not sure right now how much work would be required to get something like a responsive site output so that the docs could be called up on a mobile device, for example, but also having a desktop site that is capable of delivering a rich experience.

I'm not sure how important this is, either. With a tool that allowed manipulation of the HTML templates (and associated CSS), we should be able to add this later, but a lot will depend on the level of allowed customization. I am a big fan of CSS-based responsive layouts, and we do a little bit of that today, so that's where my first inclination would land.

> How important is it that the generation process includes good integration with IDE tools (VS, VSCode, ... Rider? Eclipse?)?

Somewhere between "not at all" and "something only you or I (or anybody who needs to modify the generator) cares about". My expectation is that the tool will be run as part of the build script (meaning, we'd need to add a target for it in `/tools/builder`). I use Visual Studio almost exclusively for day-to-day development, and I use the command line build system for pretty much everything except running/debugging individual tests. I expect our best option is to have the CI process build the HTML into a folder under `/artifacts` (we have `/artifacts/test` and `/artifacts/packages` already today). We'll use GitHub Actions to pack it up like we do the others so that it's a downloadable artifact. When I ship new versions, I can download the ZIP file and put it into the right place in the site. We may want to make some simple tasks available to us while iterating on the documentation so that recently locally built documentation can be easily placed into the appropriate place of a clone of the `gh-pages` branch for verification. (Today I use VS Code + `bundle exec jekyll serve -w` when working on the web site.)

As we discussed in e-mail, there will also need to be some "meta" work here as well so that we can provide access to multiple versions of the API documentation that would align with the releases. I'll probably want to go back and back-fill several versions, given the automation, though doing that by hand will be a task I'll take on myself, since we can't rely on the CI process in the past to do it. 😄 I don't know how extensive this needs to be, and if the generator already contains logic to handle this kind of thing, so much the better. I do think it's important for users to be able to provide exact links to documentation pages for specific versions, so I'd like to stay away from anything which swaps HTML out from underneath people when they shift to a different version without also changing the URL (that is, the version should be part of the URL).

**Reply by AlanMcBee** [reply]
Thank you for the welcome!

I think everything you described as requirements/limitation is something that I would have either expected or pursued, anyway. I feel like we're going to work well together.

I totally, 100% get the need to support prior versions. I have a wishlist I brainstormed of (potentially) useful features around docs (multiple version support is one of many), and without over-engineering any solution, I'm going to put in effort to make it reasonable to consider most of those features in future work.

**Reply by bradwilson** [reply]
We can definitely do the work in stages. I think focusing in on the mechanical conversion of XML to HTML is probably the top priority. Being able to get just the current version up onto the site is a win for sure.

The second most important thing for me at that point is whether the generation can/will include some of the version selection mechanism. The simplest "version selection mechanism" is just that every version is stand alone and we offer a page to link into the docs for specific versions.

**Reply by AlanMcBee** [reply]
I'm wondering if the part of getting releases marshalled and ready for consumption that is largely based on manual steps you (Brad) have to repeat with every release is something that we could put into automation, to help streamline the release process and free up your time.

**Reply by bradwilson** [reply]
The way I see it there are two manual steps that prevent full out automation.

The first is that I hand-write the release notes based on reviewing the commits (I don't write a note for every commit, either, since I don't restrict my personal pushes to go via a rebase'd PR, and the notes are rarely just copy/paste of the commit message). This could be semi-automated (generate the release notes and then I hand-edit them for clarity).

The second is that I like to do a final validation pass on the RTM build packages before they're pushed to NuGet, using the integration test project. The integration project isn't anything that's automated; I just pull new package versions, make sure everything is working as expected. A related step here is that any updated/new analyzers need to get new examples written in the integration project (so that I can test detection and fixes), which also means updating or adding documentation pages for those analyzers.

I honestly have a hard time envisioning fully automating all of this in a way that I'd be happy with (and that the community would be happy with), though I'm open to suggestions.

**Reply by AlanMcBee** [reply]
I read your notes on PR #2916. This is my more-complete response/thoughts for both this discussion and the current thread on the PR.

I don't advocate 100% hands-free automation, either. Even a mostly automated solution would benefit from at least manual approval steps. Mostly I'm just trying to pull into automation anything that you've been doing manually that might take you more time to put into automation than it takes to do manually.

So, I have a kind of master plan in the back of my head. I'm holding it back because you are the maintainer of this repo and know it far better than I do. But when I said I could put on the Documentation Person Hat, I meant that I'd like to work with you so that you feel comfortable delegating much (not all) of the responsibility for getting xUnit documented under your oversight. I expect that will take a while.

For the moment, I'm taking a step-by-step approach with you and the repo community. The PR I submitted is just a first step, mostly just to produce an artifact for the docs, but not to get it included as part of the xunit.net site, yet. It was intended just to prove out the DocFX step and see what else needed to be done.

So, here's a _very_ soft version of the first sequence of tasks in my back-of-the-head plan.

1. Introduce the DocFX build step to the CI workflow. That's what I started with the PR. I expect to continue with more PRs to advance the build steps so that more of the process is automated as more of these tasks are reached.
2. Convert the current documentation on `gh-pages` to markdown. Keep all the content as-is; only change it from markup (HTML) to markdown.
3. Move the converted markdown documentation from the `gh-pages` branch to a path (TBD) on the `main` branch tree. My PR introduces a new path at `/.docs/`, so maybe either put the converted markdown pages there, or at `/src/docs/`, whichever feels right.
4. Modify the DocFX config so that it merges these converted manual documentation pages with the API pages to produce a complete set of documentation HTML pages.
6. Modify the CI workflow so that it only compiles the DocFX metadata, while the PR workflow does the conversion from metadata to HTML and PDF.
7. Add a build step to the PR workflow to publish the completed documentation to the `gh-pages` branch, and whatever else needs to be done so that it's all available at xunit.net. 

I can do all these tasks. The conversion (task 2) is probably the most time-consuming. I'm happy to do that work. I will merely convert content; I won't edit (except for obvious typos).

What would this change for you? I don't know everything, but here's what I'd expect.

1. Going forward, you could keep draft of upcoming release notes as markdown in the `main` branch.
2. Documentation, like code, would be continually compiled but not published with each submitted PR. An approved PR would publish the documentation to xunit.net. We can easily omit the draft of the upcoming release notes from that.

If needed, we could (and should) modify the PR workflow so that it only publishes to `gh-pages` when you are releasing a new version.

I didn't talk about the prior versions of xUnit. I have some questions for you about that. I observed that there is a v2 of the API that is still being built and published. The docs for that are included in the current run from the PR. If you open the docs.zip artifact (the PR has a link to the one my repo built), you can see the contents of the `api` folder and scan the `toc.html` file to see what's there. But if that's not correct, and what we need to do is build the docs from the `v2` branch or one or more of the `v2.*` branches, let me know. I am sure it can be done. But do we need to hold up the current API docs until we get that solved?

I have a lot more that I feel would be valuable on the documentation front, but I wanted to just get this first step done before making plans beyond. I'll add another note to this discussion to capture my brain dump, although I think the proper next step (after we get where you are happy with the process and the product) would be to triage and work on the open doc issues submitted to the repo.

**Reply by bradwilson** [reply]
I'm going to make some comments on individual thoughts, and then I'll take a look at your proposed task list in whole.

> I observed that there is a v2 of the API that is still being built and published.

Yes, this work will need to be done in both the `v2` branch for v2, and the `main` branch for v3. In addition, we'll need a "top level" API documentation landing page that directs people to choose either v2 or v3, because the structure and content of those things will be different (assembly names are different, and in some cases things will have moved namespaces in the transition), so a "drop down" style version selector would not be appropriate across v2 and v3 (though it would be appropriate for published versions inside of each (for example, having a `2.7.0` vs. `2.7.1` vs. `2.8.0` version choice).

Back-filling the documentation metadata for these things will have to be something that's done by hand one time, since the source trees of the time won't have access to the changes that we're proposing to add to `main` and `v2`. I don't particularly see this as an onerous task, as we can just zip up and bring along the .docs folder and then run the `docfx metadata` command by hand as needed.

> Convert the current documentation on gh-pages to markdown. Keep all the content as-is; only change it from markup (HTML) to markdown.

The work here, in my mind, would be more than just converting from HTML to Markdown (which should be fairly straightforward, if mundane/mechanical), but also the conversion of the automation we use through Jekyll today to auto-generate pages.

A good example of this is the analyzer rules: today, they are placed in the [_rules folder](https://github.com/xunit/xunit/tree/gh-pages/xunit.analyzers/_rules) with header metadata so that they can be filtered and looped to create [the index page](https://github.com/xunit/xunit/blob/gh-pages/xunit.analyzers/rules/index.html). This feels very much like a scenario where we could use DocFX YAML (like toc.yml) so that it can not only generate an index page, but also a left nav like the API pages have that can jump between rules. That YAML might be hand constructors, or might be something we could automate during the site build before calling DocFX (in much the same way that `docfx metadata` is done separately from `docfx build`).

Although we don't do it today, we can use our own Actions builds for web site content (the built-in Actions system for GH Pages just runs Jekyll). I feel like being able to modify and tweak the site in the context of the `gh-pages` branch rather than trying to do it as part of the `main` or `v2` branches feels more natural to me. I would bring the C#-based build system into the `gh-pages` branch so that there is a convenient `./build` command available there as well to re-generate content based on YAML/MD changes. It also addresses the split between the `main` and `v2` branch in the sense that it would be rather awkward to change the `main` branch in order to get the `v2` branch's latest release APIs posted, if the actual site build took place in the CI process for the `main` branch. Leaving just the metadata generation in CI in those branches feels like the best division of labor to me, so the "building" of the site content still lives in the `gh-pages` branch.

An alternative would be to stop using branches entirely, and setup in essence a "monorepo" that consists of a `v2` and `v3` folder at the top level, and then things that live outside of that, like `site` for the website content, wouldn't be duplicated across branches. This would be a potentially larger (and more disruptive, for people who have clones of the repo) operation that I think would be much lower down on the priority list. It has downsides (as the single branch churns all the time as work splits between `v2` and `v3`, whereas today things happening in another branch are not disruptive), so it's also something I wouldn't undertake lightly; I'd want some experience with the system working with the existing branches before I'd be willing to experiment with a changed like that.

There are always downsides to the documentation living "apart" from the code in a different branch, but honestly in the existing "two primary branches" workflow that I'm living in right now, I don't see any other way as being realistically feasible. My mind may also change when it's time to retire v2, once v3 is finished and generally adopted.

Wherever the docs live, they'll always be "separate" from the analyzer source and Visual Studio adapter source, because those live in separate repositories today. So updating documentation related to those two things always involves the separate repo/branch for the workflow, unless part of the proposed monorepo would be to also bring the analyzers and VS adapter into the main repo as top-level folders.

> Modify the CI workflow so that it only compiles the DocFX metadata, while the PR workflow does the conversion from metadata to HTML and PDF.

I personally would not be reviewing HTML or PDF for every PR; reviewing the source code is sufficient to ensure that things are appropriately XML Doc Commented for the APIs. I would be content with the CI builds just building the metadata (whether RTM or now) and including that in the downloadable artifacts, as that means I could pull the metadata and push it into whatever the correct folder structure is for the site and then just build the docs and see what it looks like.

> I can do all these tasks.

There are some things it would make more sense for me to do, like bringing the build system into `gh-pages` (perhaps at first just mirroring the existing Pages build via Jekyll) so that we can then move over to a DocFX build (even prior to the API integration) as a next step. Then as you do step 2 above you can easily build the site as needed and browse it. I'm convinced DocFX is capable of doing the hardest part of the problem (the API doc generation) and I'm 99.9% sure that it can be made to do what little extra automation we need for the existing mostly static content. It might make sense for that to the be the next first focus as being an "aside" from the API stuff, which we'd be able to integrate later.

> So, here's a very soft version of the first sequence of tasks in my back-of-the-head plan.
> 1. Introduce the DocFX build step to the CI workflow. That's what I started with the PR. I expect to continue with more PRs to advance the build steps so that more of the process is automated as more of these tasks are reached.
> 2. Convert the current documentation on gh-pages to markdown. Keep all the content as-is; only change it from markup (HTML) to markdown.
> 3. Move the converted markdown documentation from the gh-pages branch to a path (TBD) on the main branch tree. My PR introduces a new path at /.docs/, so maybe either put the converted markdown pages there, or at /src/docs/, whichever feels right.
> 4. Modify the DocFX config so that it merges these converted manual documentation pages with the API pages to produce a complete set of documentation HTML pages.
> 5. Modify the CI workflow so that it only compiles the DocFX metadata, while the PR workflow does the conversion from metadata to HTML and PDF.
> 6. Add a build step to the PR workflow to publish the completed documentation to the gh-pages branch, and whatever else needs to be done so that it's all available at xunit.net.

Here is what I propose are the first next steps:

1. **(Brad)** Convert to a custom Actions build for the `gh-pages` branch, utilizing Jekyll. Bring in the C#-based build system so that building and viewing the content is straightforward (build via `jekyll`, view via Docker container running nginx pointed at the generated content folder). I would like to investigate if there is an easy way with VS Code to re-run the site build when it detects filesystem changes so that we can preserve the "save you changes, wait a few seconds, then F5 in the browser to see the changes" workflow that has been implicit by using `jekyll serve`. _This can be pushed as soon as it's ready._
1. In a branch off of `gh-pages`:
    1. **(Brad)** Convert the build from Jekyll to DocFX, understanding that with the existing content this will break things which rely on Jekyll page generation and/or anything which is HTML instead of Markdown.
    1. **(Alan)** Convert the current documentation on `gh-pages` to Markdown. Keep all the content as-is; only change it from markup (HTML) to Markdown.
    1. **(Whoever gets to it)** Convert the existing Jekyll-generated content to toc.yml (or whatever is most appropriate) so that DocFX can generate that content for us.
    1. **(Both)** Review the navigation and structure of the site to see what works best for DocFX and for user flow. See where we can utilize DocFX features that wouldn't have been available in Jekyll, like having the API-style left nav available in generated content places. This would definitely include the analyzers documentation, but we may be able to find other places where it makes sense too (like the structure of the release notes).

I think after Step 2 above, we'll have a significant step forward in the documentation process, and then we can start to analyze how best to integrate in the APIs so we can get the page/URL structure we want, navigation we want, etc.

What do you think?

**Reply by bradwilson** [reply]
I should also mention: I have done all my work with `gh-pages` in WSL because getting Ruby and whatnot running there is trivial for me (with my [`ansible-dev-pc`](https://github.com/bradwilson/ansible-dev-pc) scripts) whereas I've found Ruby on Windows to be rather problematic (and can generate different results than using Linux). With the move to DocFX, I'm open to ensuring that this whole process works in either Windows or Linux (or both); assuming the tooling is cooperative, I would be good with moving my personal process to Windows (or keeping it on Linux).

**Reply by bradwilson** [reply]
> Convert to a custom Actions build for the gh-pages branch

I have done this and pushed it to the `gh-pages` branch. To clean up the folder structure, I moved all the site content under `/site` (the generated content will still go to `/_site` as it did previously.

I've added two build targets:

- `./build` (or `./build build`) will build the site from the Jekyll templates into `/_site`
- `./build serve` will run the Jekyll server which live recompiles files when the change

Everything else should just work as previous, except that now we're using our custom Actions workflow instead of the built-in workflow, so that we can easily convert away from Ruby/Jekyll when the time is right.

**_IMPORTANT:_** Because I moved all the content under `/site`, you'll want to pull before making any changes. Everything was recorded as a git rename, so it should take any committed edits and apply them appropriately, but merge issues might crop up, so the sooner the better. 😄 

As pre-requisites, you'll want to have Ruby 2.7.4 and Ruby Bundler 2.1.4 installed. I have tested this on both Linux and Windows and it does work in both places.

**Reply by AlanMcBee** [reply]
Whew! Okay, I think I'm getting a clearer picture of your current constraints and goals. I have no problem working from your outline of the tasks.

Would you like to kill the PR I submitted, or have me cancel it myself? Not sure if it matters to you.

I am pulling your changes now and will set up my build environment. I'm assuming you're working on 2(i) "convert build from jekyll to DocFX". LMK if I can help since I've been doing some deep dives on DocFX.

**Reply by bradwilson** [reply]
No need to cancel it yet, I've been iterating on it locally even though it's a couple steps down the road.

**Reply by bradwilson** [reply]
I will be working on 2(i) next, though that's going to get moved to a new branch. The first version of the DocFX build system will just use the existing content in advance of adding the API docs later.

Which would you prefer?

(a) work on the aforementioned new branch so that the HTML => Markdown conversions you're working on are _also_ getting things ready for DocFX; or
(b) work on the `gh-pages` branch so that the HTML => Markdown conversions are based on the existing Jekyll templates, and then I/you/we could do the move to DocFX separately?

To me, (a) feels like combining at least 2(ii) and 2(iii) into a single step, whereas (b) is just step 2(ii). As such, I had been assuming (b) (also on the theory that smaller steps are easier), but if you think (a) is better then I will prioritize getting that new branch set up ASAP.

**Reply by bradwilson** [reply]
BTW the reason I planned to do 2(i) on the separate branch is because we won't be ready to merge the move to DocFX until at least 2(iii) is done, if not also waiting for 2(iv). If we do 2(ii) separately from 2(iii), then merging it into `gh-pages` when it's ready to use Jekyll seems like the right next step.

**Reply by AlanMcBee** [reply]
Well, like I said at the outset, I'm totally fine doing anything and everything that is related to documentation. I'm encouraging you to delegate stuff to me. I'll make sure that you understand and agree with anything I'm doing. So I can just start doing all of step 2 (i through iii) so that we can get to 2(iv). I'm in a good place to put in plenty of time just now. So I'm ready to do (a) and then (b), including any work that is needed to keep the new branch up to date with anything that might happen on the `gh-pages` branch until we're ready to merge the new branch in. Is that all right with you?

**Reply by bradwilson** [reply]
Sounds good to me! Since you're doing all the work, feel free to just set up your own branch, and when you're done we can do it as a normal PR against `gh-pages`.

**Reply by bradwilson** [reply]
To implement `./build serve` (which does a watch & rebuild & serve), since DocFX does not support `docfx watch` yet, there are some suggestions in the feature request thread that are probably reasonable substitutions while we wait for the feature, like:

https://github.com/dotnet/docfx/issues/2297#issuecomment-1444721201

**Reply by bradwilson** [reply]
I'm releasing a new build today, and I decided I should stop writing things in HTML and start writing them in Markdown so I don't just grow your backlog unnecessarily. 😂 Might as well start with the release notes, right?

In an attempt to switch, it become apparent to me how much I'm doing that isn't Markdown compatible, like adding CSS classes and custom attributes to things. I spent some time noodling over the best way to completely get rid of custom HTML in the release notes, and I ended up with some creating CSS rule writing via a custom stylesheet that is only used for release notes. I considered also converting the index page, but I felt a little held back on the way tables are styled there, and I didn't want to bite off too much at once.

Would be curious on your feedback on this commit, and the general strategy of using small CSS stylesheets that get added to particular page "types" to help keep the Markdown HTML-free. This is obviously going to be an ongoing problem across the site as everything moves to Markdown. https://github.com/xunit/xunit/commit/31ba3169c5dd57e7da1ab91c7f063e2c60b7ab93

**Reply by AlanMcBee** [reply]
Sorry for the response delay. I'll review the commit right away and share my feedback later today (I think).

I'll investigate if we can do something helpful and simple with the tables.
